spec/support/controller_helpers.rb

```diff
@@ -0,0 +1,12 @@
+module ControllerHelpers
+  def login_with(user = double('user'), scope = :user)
+    current_user = "current_#{scope}".to_sym
+    if user.nil?
+      allow(request.env['warden']).to receive(authenticate!).and_throw(:warden, scope: scope)
+      allow(controller).to receive(current_user).and_return(nil)
+    else
+      allow(request.env['warden']).to receive(:authenticate!).and_return(user)
+      allow(controller).to receive(current_user).and_return(user)
+    end
+  end
+end
```

Comment:
Typically I just stub controller.current_user. I've never had to stub or deal with warden specifically.